### PR TITLE
fix ZeroDivisionError and epoch counting

### PR DESCRIPTION
**Pull request by ydshieh**
@sgugger 

This PR fix 2 minor bugs in `trainer.py`.

First, the two lines
            
    num_train_epochs = (
        self.args.max_steps // (len(train_dataloader) // self.args.gradient_accumulation_steps) + 1
    )

and

    epochs_trained = self.global_step // (len(train_dataloader) // self.args.gradient_accumulation_steps)

gives `ZeroDivisionError` when `len(train_dataloader) < self.args.gradient_accumulation_steps`.

The fix takes into account the code below the comment

     # last step in epoch but step is always smaller than gradient_accumulation_steps

therefore when `len(train_dataloader) < self.args.gradient_accumulation_steps`, we still have 1 step in each epoch.

The second bug is due to the `+ 1` in the

    num_train_epochs = (
        self.args.max_steps // (len(train_dataloader) // self.args.gradient_accumulation_steps) + 1
    )

In the example below, we have

    len(train_dataloader) == 40
    self.args.gradient_accumulation_steps = 4
    self.args.max_steps = 10

However we get `num_train_epochs == 2` from the original code, which should be `1`.

    python utils/download_glue_data.py --data_dir ./examples/text-classification/glue/ --tasks all

    python3 run_glue.py \
        --task_name wnli \
        --data_dir ./glue/WNLI \
        --model_name_or_path distilbert-base-uncased \
        --output_dir ./glue/WNLI/ \
        --max_seq_length  16 \
        --num_train_epochs 1 \
        --per_device_train_batch_size 16 \
        --gradient_accumulation_steps 4 \
        --max_steps 10 \
        --logging_steps 1 \
        --save_steps 5 \
        --seed 1 \
        --do_train \
        --do_eval \
        --do_predict \
        --overwrite_output_dir

    (Pdb) len(train_dataloader)
    40
    (Pdb) self.args.gradient_accumulation_steps
    4
    (Pdb) self.args.max_steps
    10
    (Pdb) p num_train_epochs
    2

### CODE REVIEW

**Comment by sgugger** [comment]
Looks great! Would you mind adding a few tests so we don't accidentally make changes that will resurface this bug?

**Comment by chiapas** [comment]
> 
> 
> Looks great! Would you mind adding a few tests so we don't accidentally make changes that will resurface this bug?

I did a few simple tests on my side. Could you guide me how to write a test that is usually done by HF members? BTW, I don't have powerful machine, so hope the tests could be lightweight.

**Comment by sgugger** [comment]
Look at the `test_trainer.py` file in the tests directory. There are utility functions to build a small `Trainer` on a simple regression problem. You can just add a few tests mimicking the ones that count the number of steps during training.

To check they pass, you can run the command:
```
pytest tests/test_trainer.py
```
which should go pretty quickly.

**Comment by chiapas** [comment]
Thanks. I will do it and report it when ready.

**Comment by chiapas** [comment]
I add a test `test_num_train_epochs_in_training` for the case `len(train_dataloader) < self.args.gradient_accumulation_steps`.
For the offset by 1 bug below due to the `+ 1` in 

    num_train_epochs = (
        self.args.max_steps // (len(train_dataloader) // self.args.gradient_accumulation_steps) + 1
    )

, I didn't figure a way to test, because this information is not returned in the output, and its actual usage is for this line

    logger.info("  Num Epochs = %d", num_train_epochs)

because the training loop is actually controlled by `self.args.max_steps` if it is given.

(I assume `adding a few tests` means to add in the `test_trainer.py` and push).

File: tests/test_trainer.py [hunk 1]
```diff
@@ -299,3 +299,19 @@ def test_trainer_iterable_dataset(self):
         trainer = Trainer(model=model, args=training_args, train_dataset=train_dataset)
         loader = trainer.get_train_dataloader()
         self.assertIsInstance(loader, torch.utils.data.DataLoader)
+
+    @require_non_multigpu
```
Comment (sgugger) [inline tests/test_trainer.py:303]: This flag won't be necessary here, as the test will work equally well on a multigpu setup.

**Comment by sgugger** [comment]
Yes adding the test in `test_trainer.py` is exactly what I wanted, thanks!

File: tests/test_trainer.py (hunk 1 above)
Comment (chiapas) [reply tests/test_trainer.py:303]: OK. I kind copy paste `test_number_of_steps_in_training` which has `@require_non_multigpu`, since I am not familar with pytorch when it comes with distributed stuffs. I will remove.

**Review by LysandreJik — APPROVED** [review]
Great, neat change and thanks for adding a test!
